### ISSUE TITLE
Cache class creation to improve memory usage and performance

### ISSUE THREAD

**Issue by mattz0rt** [Vector35/binaryninja-api#1120]
While running analysis on larger binaries we've encountered issues with binja taking up multiple GBs of memory due to new classes being created for each call to various accessors for asm/llil/mlil functions/blocks/instructions. We've seen a significant reduction (50%+) of memory usage by caching these accesses as well as performance improvements, and were wondering if it would make sense to move the caching into the python api as a configurable feature (so we don't need to add wrappers around all the various calls).

This could easily be done by using a cached class decorator ([ref](https://github.com/ActiveState/code/blob/3b27230f418b714bc9a0f897cb8ea189c3515e99/recipes/Python/577998_Cached_Class/recipe-577998.py)) for the various data classes in function.py, lowlevelil.py, and mediumlevelil.py (and perhaps others).

Would be happy to do a PR if you guys think this makes sense to add, caveat being that I am just starting to get familiar with the internals.

**Comment by plafosse** [comment]
Sorry it's been so long to get to responding to this issue. We do agree that a cached class decorator would work. We'd be happy to accept a PR for this issues, as its not a high priority for us.

**Comment by rssor** [comment]
In terms of the IL instructions, caching is likely to result in potentially unexpected memory use blowup, as those objects hold references to the IL function objects and thus keep the extended analysis information in the core alive until they are garbage collected. You'd very much be at the whims of the caching decorator used and the function/il instruction access pattern as to whether or not it ballooned memory use or not. (For example, an analysis pass that looked at one instruction from a thousand different functions would potentially keep the extended analysis data for all one thousand functions alive until they aged out, and this would be enormously expensive)

On a related note, though, we could have huge efficiency wins by being lazy about parsing out the instruction operands. Right now getting an IL instruction will extract the entire expression tree under it, including all of the operation-dependent operand extraction, which is work we could get away with deferring.

Do you have any finer-grained info on which classes were the biggest offenders? I see some obvious sources of inefficiency for mlil/llil instruction objects, and generally speaking the intention is for things like functions/basic blocks objects to be cheap to just duplicate as those are just references to an object inside the core (which is not duplicated).

**Comment by mattz0rt** [comment]
A lot has changed since this was first proposed, and we decided to just keep the caches that seemed to result in performance gains and not worry too much about memory usage (since that's easier to increase vs cpu speeds). This probably wouldn't make as much sense for others since it's very much based on our specific use cases, but here are the functions that we currently cache:
- MediumLevelILInstruction.ssa_form
- MediumLevelILFunction.\_\_getitem\_\_
- Function.symbol

We also ended up creating a separate layer between the binja python api and our code to add these caches (instead of modifying the python api itself) and it's a bit hacky to make it work without modifying binja code, so it's probably not a good fit for PR at this point. 

Let me know if there's any additional information that would be helpful to know. Thanks!
